selfservice/flow/saml/handler.go

```diff
@@ -0,0 +1,325 @@
+}
+
+func (h *Handler) RegisterPublicRoutes(router *x.RouterPublic) {
+
+	h.d.CSRFHandler().IgnorePath(RouteSamlLoginInit)
+	h.d.CSRFHandler().IgnorePath(RouteSamlAcs)
+
+	router.GET(RouteSamlMetadata, h.serveMetadata)
+	router.GET(RouteSamlLoginInit, h.loginWithIdp)
+}
+
+// Handle /selfservice/methods/saml/metadata
+func (h *Handler) serveMetadata(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	config := h.d.Config(r.Context())
+	if samlMiddleware == nil {
+		if err := h.instantiateMiddleware(*config); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		}
+	}
+
+	samlMiddleware.ServeMetadata(w, r)
+}
+
+// swagger:route GET /self-service/methods/saml/auth v0alpha2 initializeSelfServiceSamlFlowForBrowsers
+//
+// Initialize Authentication Flow for SAML (Either the login or the register)
+//
+// If you already have a session, it will redirect you to the main page.
+//
+//     Schemes: http, https
+//
+//     Responses:
+//       200: selfServiceRegistrationFlow
+//       400: jsonError
+//       500: jsonError
+func (h *Handler) loginWithIdp(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+
+	// Middleware is a singleton so we have to verify that it exist
+	if samlMiddleware == nil {
+		config := h.d.Config(r.Context())
+		if err := h.instantiateMiddleware(*config); err != nil {
+			h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+		}
+	}
+
+	conf := h.d.Config(r.Context())
+
+	// Checks if the user already have an active session
+	if e := new(session.ErrNoActiveSessionFound); errors.As(e, &e) {
+		// No session exists yet, we start the auth flow and create the session
+		samlMiddleware.HandleStartAuthFlow(w, r)
+	} else {
+		// A session already exist, we redirect to the main page
+		http.Redirect(w, r, conf.SelfServiceBrowserDefaultReturnTo().Path, http.StatusTemporaryRedirect)
+	}
+}
+
+func DestroyMiddlewareIfExists() {
+	if samlMiddleware != nil {
+		samlMiddleware = nil
+	}
+}
+
+func (h *Handler) instantiateMiddleware(config config.Config) error {
+	// Create a SAMLProvider object from the config file
+	var c samlstrategy.ConfigurationCollection
+	conf := config.SelfServiceStrategy("saml").Config
+	if err := jsonx.
+		NewStrictDecoder(bytes.NewBuffer(conf)).
+		Decode(&c); err != nil {
+		return errors.Wrapf(err, "Unable to decode config %v", string(conf))
+	}
+
+	// Key pair to encrypt and sign SAML requests
+	keyPair, err := tls.LoadX509KeyPair(strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PublicCertPath, "file://", "", 1), strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PrivateKeyPath, "file://", "", 1))
+	if err != nil {
+		return err
+	}
+	keyPair.Leaf, err = x509.ParseCertificate(keyPair.Certificate[0])
+	if err != nil {
+		return err
+	}
+
+	var idpMetadata *samlidp.EntityDescriptor
+
+	// We check if the metadata file is provided
+	if c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"] != "" {
```

Comment:
I feel like this access lacks safety railings:

 * if len(c.SAMLProviders) == 0 this method will panic. Example at https://go.dev/play/p/AAD95wCtX_P
 * Is IDPInformation always initialized ?
 * Is idp_metadata_url always set ? Aren't there cases when it is empty but still valid ?

```suggestion
	if len(c.SAMLProviders) > 0 {
	     if val, ok := c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"]; ok && len(val) > 0 {
```